### ISSUE TITLE
Cannot Switch to Different Python kernel

### ISSUE THREAD

**Issue by nathanlesage** [microsoft/vscode-jupyter#5214]
## Environment data

-   VS Code version: Version: 1.54.3 (Universal) Commit: 2b9aebd5354a3629c3aba0a5f5df49f43d6689f8
-   Jupyter Extension version (available under the Extensions sidebar): v2021.3.619093157
-   Python Extension version (available under the Extensions sidebar): v2021.3.658691958
-   OS (Windows | Mac | Linux distro) and version: macOS Big Sur 11.2.3 (ARM 64 arch)
-   Python and/or Anaconda version: 3.8.2, conda 4.9.2
-   Type of virtual environment used (N/A | venv | virtualenv | conda | ...): conda
-   Jupyter server running: Local

## Expected behaviour

When selecting a specific Python Kernel (which are found) Jupyter VS Code should use that one, at least after a restart. In my workspace settings, I have explicitly set the option `"python.pythonPath": "/opt/homebrew/Caskroom/miniforge/base/envs/nlpsimple/bin/python"`, where `nlpsimple` is my desired environment.

## Actual behaviour

Running `sys.executable` outputs `'/opt/homebrew/Caskroom/miniforge/base/bin/python'`, so apparently VS Code Jupyter is unable to switch to a different environment. I have switched to my desired environment in the internal terminal before re-starting the kernel and re-running this command with the same result.

**It seems that Jupyter VS Code tries to use my system installation of Python to start the IPython Kernel. However, my environments use 3.8 since I need PyTorch, while my system installation has 3.9 -- see logs below.**

## Steps to reproduce:

[**NOTE**: Self-contained, minimal reproducing code samples are **extremely** helpful and will expedite addressing your issue]

1. I am using Jupyter VS Code in a pretty fresh install since VS Code just recently released as stable for M1 Macs. Other than the above-mentioned setting in my workspace, there is nothing I changed with the base settings, and I have not yet worked using VS Code Jupyter in this new installation (which should be self-contained from the Insiders edition I was using before)
2. The environments were installed before, however.

## Logs

<details>

<summary>Output for <code>Jupyter</code> in the <code>Output</code> panel (<code>View</code>→<code>Output</code>, change the drop-down the upper-right of the <code>Output</code> panel to <code>Jupyter</code>)
</summary>

<p>

```
[I 09:51:42.430 NotebookApp] Kernel shutdown: a9c72ff0-b1eb-44ee-b83e-a3ba17234563
> /opt/homebrew/Caskroom/miniforge/base/envs/nlpsimple/bin/python ~/.vscode/extensions/ms-toolsai.jupyter-2021.3.619093157/pythonFiles/pyvsc-run-isolated.py -c "import ipykernel"
[I 09:51:43.013 NotebookApp] Kernel started: 4f1478ad-4875-4935-b196-5dd89c57c141, name: python3
[W 09:51:43.017 NotebookApp] delete /t-20bc9970-c10a-498b-9b6a-b8f48d942aff.ipynb
Error 2021-03-19 09:54:42: KernelSpec has interpreter information, however a matching interpreter could not be found for /opt/homebrew/opt/python@3.9/bin/python3.9
Error 2021-03-19 09:54:43: KernelSpec has path information, however a matching interpreter could not be found for /opt/homebrew/opt/python@3.9/bin/python3.9
Error 2021-03-19 09:59:19: KernelSpec has interpreter information, however a matching interpreter could not be found for /opt/homebrew/opt/python@3.9/bin/python3.9
> /opt/homebrew/Caskroom/miniforge/base/envs/nlpsimple/bin/python ~/.vscode/extensions/ms-toolsai.jupyter-2021.3.619093157/pythonFiles/pyvsc-run-isolated.py -c "import ipykernel"
> /opt/homebrew/Caskroom/miniforge/base/envs/nlpsimple/bin/python ~/.vscode/extensions/ms-toolsai.jupyter-2021.3.619093157/pythonFiles/pyvsc-run-isolated.py pip list
Error 2021-03-19 09:59:20: KernelSpec has path information, however a matching interpreter could not be found for /opt/homebrew/opt/python@3.9/bin/python3.9
> /opt/homebrew/Caskroom/miniforge/base/envs/nlpsimple/bin/python ~/.vscode/extensions/ms-toolsai.jupyter-2021.3.619093157/pythonFiles/pyvsc-run-isolated.py pip list
```

</p>
</details>

**Comment by njbamboo** [comment]
I have the same problem. The only way to solve it is to change the jupyter server from the vscode default to the link given by  jupyter notebook in cmd/powershell (I opened another jupyter notebook by command). The link is like "http://localhost:8888/?token=xxxxxxxx  or http://127.0.0.1:8888/?token=xxxxxxx".  When I changed the jupyter server to this one, I can import every module. Also I found a strange thing : by changing the jupyter serve between the above two choices, when you switch back to vscode's default server at some time point (seem to be random), it will be possible to import the module again!

**Comment by DonJayamanne** [comment]
Thanks for filing this issue. I believe we have addressed this in our current development version of the extension.

In order to validate the fixes please could you:
* Install VS Code Insiders
* Install the Jupyter extension
Test and let me know if this works.

Thanks

**Comment by nathanlesage** [comment]
Thank you for the swift response! I thought I was able to get rid of Code Insiders!

Just downloaded it, and installed the necessary extensions, but the same problem ensued because of which I was happy to get rid of Code insiders in the first place: Nothing works, I cannot start a Kernel there, no matter what I do. So unfortunately, I cannot check the changes …

**Comment by DonJayamanne** [comment]
@nathanlesage 
We've made a lot of fixes related to this in the current development version.
I'm sorry it didn't work
Please could you go through the trouble of installing it again and provideing the logs.
Please could you enable logging as follows:
* Open settings & go into `Jupyter->Logging`
* Change the value to `debug`
* Reload VS Code Insiders,
* Attempt to repro the issue & then copy the output from the `Jupyter` output panel.

The code in our development branch has diverged a lot compared to what was released last, and it will be difficult to identify the cause, as we've fixed a lot of similar issues (if not identical).
hence the request for the logs from insiders, we may have missed something.

Thanks.

**Comment by muser840** [comment]
Just to confirm that Jupyter uses the wrong python, which is confirmed by
``` python
import sys
print(sys.executable)
! which python
```

For me the output of the above prints:
```
/usr/bin/python3
/home/user/venv/bin/python
```

It should be the same! These is some mess with switching between python versions by the extension. Somehow switching to old and already deleted environments make it work. That makes extension totally unusable...

The problem is present on versions:
v2021.5.702919634
v2021.6.732501275

**Comment by DonJayamanne** [comment]
@muser840 Thanks for the information.
I'd like to get to the bottom of this, please could you create a new issue with the lots from the Jupyter panel.

* Attempt to repro the issue & then copy the output from the `Jupyter` output panel.

I'd like to help you get unblocked (I'm certain there are others who run into the same issue) and would like to ensure we look at this separately for now.

**Comment by DonJayamanne** [comment]
Closing this issue due to inactivity.
If this issue still exists, please do get back to us or please feel free to file a new issue

@muser840 Please do create a new issue, I'll help you get unblocked there.
